requirements/listing-api-requirements.md

```diff
@@ -0,0 +1,40 @@
+# Listing API Requirements
+
+Registries provide the basic means to push, pull, discover and manage content. While push and pull are generally feature-rich and standardized, the discover and manage APIs are extremely limited. Each registry project and operator have had to implement discover (list) and manage APIs, and each of them are unique to that particular registry.
+
+To foster consistency across registries, enabling the community to build tooling that works consistently across all [OCI distribution-spec conformant registries][oci-compliance], we propose a common set of listing APIs.
+
+To facilitate discussions, prioritization and phased implementations, we will capture a list of scenarios and requirements. Once the requirements are felt complete and prioritized, we can transition to a design that implements the agreed prioritized items, while reserving space for the future enhancements.
+
+For the purposes of examples within a scenario, we'll assume an `oci-reg` cli exists. The purpose is not to design the specific CLI, rather provide concrete examples whose specific names and parameters will evolve.
+
+## Scenarios
+
+1. A user can get a list of repositories within a given registry, within a specified org and/or namespace.  
+      - `oci-reg list acme-rockets.io/`
+      - `oci-reg list acme-rockets.io/org1/`
+2. A user can get a list of tags within a given registry/namespace.
```

Comment:
It would be useful if the list could (optionally) not only contain the tags name (as-is) but also other useful details such as the underlying manifest digest and media type, and possibly the tag creation timestamp.

It is common to need such details (e.g., when showing a tag list in a UI). Looping over a name list and doing multiple requests (one for the manifest, another for the configuration, if any) requires too much time and resources.

Reply:
@joadrp, this is a great topic. I've created another [PR for the show/get-info requirements](https://github.com/opencontainers/distribution-spec/pull/232). I split them out to try and get a consensus on one at a time. But, your point of having to loop through a list to get details on each may be problematic as well. Open to ideas on how we split or merge a subset?

Reply:
> I've created another PR for the show/get-info requirements. I split them out to try and get a consensus on one at a time.

@SteveLasker, thanks. I'll have a look at it and share my opinion there.

> But, your point of having to loop through a list to get details on each may be problematic as well

I should clarify: Some of the API features that I'm looking forward to are not feasible to implement on registries that rely on filesystem metadata (like the distribution registry does), at least not in a way that would allow them to be usable. For example, this is why the distribution registry doesn't support pagination on the tag list endpoint.

Doing a recursive walk on a remote blob storage backend and inspecting file by file is not only slow but expensive. This is especially true for large repositories, such as those with thousands of tags, where a single tag list request (with no additional details, just the tag names) can take many seconds to complete.

At some point, these limitations become a major blocker for providers. They can't ask users to wait this much time to see a list of tags in the UI (or get it through the API), nor can they fulfill most feature requests because the implementation would be extremely slow and costly.

As a result, most major registries have to adopt some database to store metadata, such as Harbor with PostgreSQL and Docker Trusted Registry with RethinkDB. This enables them to use fast and flexible query languages and attach additional metadata to each artifact, metadata which is not part of the image spec (e.g., image _push_ date, push/pull count, repository size, etc.). And this is when they start implementing their own custom APIs.

This brings me to a key question: Is the intention to limit the distribution API spec to the information contained, e.g., in image manifests? Or do we want to look beyond that?